server/src/main/java/org/folio/shared/index/api/OaiService.java

```diff
@@ -0,0 +1,374 @@
+    f852.clear();
+    f852.addAll(holdings);
+    String xmlMetadata = XmlJsonUtil.convertJsonToMarcXml(marc);
+    return "    <metadata>\n" + xmlMetadata + "\n    </metadata>\n";
+  }
+
+  static Future<String> getXmlRecordMetadata(Storage storage, SqlConnection conn, UUID clusterId,
+      List<String> matchValues) {
+    String q = "SELECT * FROM " + storage.getBibRecordTable()
+        + " LEFT JOIN " + storage.getClusterRecordTable() + " ON record_id = id "
+        + " WHERE cluster_id = $1";
+    return conn.preparedQuery(q)
+        .execute(Tuple.of(clusterId))
+        .map(rowSet -> getMetadata(rowSet, clusterId, matchValues));
+  }
+
+  static Future<List<String>> getClusterValues(Storage storage, SqlConnection conn,
+      UUID clusterId) {
+    return conn.preparedQuery("SELECT match_value FROM " + storage.getClusterValuesTable()
+            + " WHERE cluster_id = $1")
+        .execute(Tuple.of(clusterId))
+        .map(rowSet -> {
+          List<String> values = new ArrayList<>();
+          rowSet.forEach(row -> values.add(row.getString("match_value")));
+          return values;
+        });
+  }
+
+  static Future<String> getXmlRecord(Storage storage, SqlConnection conn, UUID clusterId,
+      LocalDateTime datestamp, String oaiSet, boolean withMetadata) {
+    Future<List<String>> clusterValues = Future.succeededFuture(Collections.emptyList());
+    if (withMetadata) {
+      clusterValues = getClusterValues(storage, conn, clusterId);
+    }
+    // When false withMetadata could optimize and not join with bibRecordTable
+    return clusterValues.compose(values -> getXmlRecordMetadata(storage, conn, clusterId,
+        values)
+        .map(metadata ->
+        "    <record>\n"
+            + "      <header" + (metadata == null ? " status=\"deleted\"" : "") + ">\n"
+            + "        <identifier>"
+            + XmlJsonUtil.encodeXmlText(encodeOaiIdentifier(clusterId)) + "</identifier>\n"
+            + "        <datestamp>"
+            + XmlJsonUtil.encodeXmlText(datestamp.atZone(ZoneOffset.UTC).toString())
+            + "</datestamp>\n"
+            + "        <setSpec>" + XmlJsonUtil.encodeXmlText(oaiSet) + "</setSpec>\n"
+            + "      </header>\n"
+            + (withMetadata && metadata != null ? metadata : "")
+            + "    </record>\n"));
+  }
+
+  static Future<Void> listRecordsResponse(RoutingContext ctx, Storage storage, SqlConnection conn,
+      String sqlQuery, Tuple tuple, boolean withMetadata) {
+    String elem = withMetadata ? "ListRecords" : "ListIdentifiers";
+    return conn.prepare(sqlQuery).compose(pq ->
+        conn.begin().compose(tx -> {
+          HttpServerResponse response = ctx.response();
+          oaiHeader(ctx, 200);
+          response.write("  <" + elem + ">\n");
+          RowStream<Row> stream = pq.createStream(100, tuple);
+          stream.handler(row -> {
+            stream.pause();
+            getXmlRecord(storage, conn,
```

Comment:
Similarly, can we either move the DB access code to the Storage/OaiStorage or use another layer between the OaiService and the Storage that's responsible for converting DB rows into XML records (e.g expose an Iterator from Storage and and pass it an OaiRecordWriter or use lambdas if that's more the vert.x style)?